syft/pkg/cataloger/cpe_candidate.go

```diff
@@ -0,0 +1,107 @@
+package cataloger
+
+import (
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/scylladb/go-set/strset"
+)
+
+// this is a static mapping of known package names (keys) to official cpe names for each package
+type candidatesByPackageType map[pkg.Type]map[string][]string
+
+type filterCpeFieldCandidateFn func(cpeFieldCandidate) bool
+
+type cpeFieldCandidate struct {
+	value                       string
+	disallowSubSelections       bool
+	disallowDelimiterVariations bool
+}
+
+type cpeFieldCandidateSet struct {
+	candidates map[cpeFieldCandidate]struct{}
+}
+
+func filterCandidatesBySubselection(c cpeFieldCandidate) bool {
+	return c.disallowSubSelections
+}
+
+func filterCandidatesByDelimiterVariations(c cpeFieldCandidate) bool {
+	return c.disallowDelimiterVariations
+}
+
+func newCPRFieldCandidateFromSets(sets ...*cpeFieldCandidateSet) *cpeFieldCandidateSet {
+	s := newCPRFieldCandidateSet()
+	for _, set := range sets {
+		s.add(set.list()...)
+	}
+	return s
+}
+
+func newCPRFieldCandidateSet(values ...string) *cpeFieldCandidateSet {
+	s := &cpeFieldCandidateSet{
+		candidates: make(map[cpeFieldCandidate]struct{}),
+	}
+	s.addValue(values...)
+	return s
+}
+
+func (s *cpeFieldCandidateSet) addValue(values ...string) {
+	for _, value := range values {
+		// default candidate as an allow-all
+		candidate := cpeFieldCandidate{
+			value: value,
+		}
+		s.candidates[candidate] = struct{}{}
+	}
+}
+
+func (s *cpeFieldCandidateSet) add(candidates ...cpeFieldCandidate) {
+	for _, candidate := range candidates {
+		s.candidates[candidate] = struct{}{}
+	}
+}
+
+func (s *cpeFieldCandidateSet) clear() {
+	s.candidates = make(map[cpeFieldCandidate]struct{})
+}
+
+func (s *cpeFieldCandidateSet) union(others ...*cpeFieldCandidateSet) {
+	for _, other := range others {
+		s.add(other.list()...)
+	}
+}
+
+func (s *cpeFieldCandidateSet) list(filters ...filterCpeFieldCandidateFn) (results []cpeFieldCandidate) {
+candidateLoop:
+	for c := range s.candidates {
+		for _, fn := range filters {
+			if fn(c) {
```

Comment:
this is backwards from what is expected (functionally) where we should keep values where the filter function returns true.

Reply:
Trying to reverse the logic made most of this not make sense (to me)... happy to talk through that though. I went through and changed the names to make it more readable though.